cpp/benchmarks/io/orc/orc_reader_input.cpp

```diff
@@ -0,0 +1,185 @@
+
+#include <nvbench/nvbench.cuh>
+
+constexpr int64_t data_size        = 512 << 20;
+constexpr cudf::size_type num_cols = 64;
+
+namespace cudf_io = cudf::io;
+
+enum class data_type : int32_t {
+  INTEGRAL  = static_cast<int32_t>(type_group_id::INTEGRAL_SIGNED),
+  FLOAT     = static_cast<int32_t>(type_group_id::FLOATING_POINT),
+  DECIMAL   = static_cast<int32_t>(type_group_id::FIXED_POINT),
+  TIMESTAMP = static_cast<int32_t>(type_group_id::TIMESTAMP),
+  STRING    = static_cast<int32_t>(cudf::type_id::STRING),
+  LIST      = static_cast<int32_t>(cudf::type_id::LIST),
+  STRUCT    = static_cast<int32_t>(cudf::type_id::STRUCT)
+};
+
+// NVBENCH_DECLARE_ENUM_TYPE_STRINGS macro must be used from global namespace scope
+NVBENCH_DECLARE_ENUM_TYPE_STRINGS(
+  data_type,
+  [](data_type value) {
+    switch (value) {
+      case data_type::INTEGRAL: return "INTEGRAL";
+      case data_type::FLOAT: return "FLOAT";
+      case data_type::DECIMAL: return "DECIMAL";
+      case data_type::TIMESTAMP: return "TIMESTAMP";
+      case data_type::STRING: return "STRING";
+      case data_type::LIST: return "LIST";
+      case data_type::STRUCT: return "STRUCT";
+      default: return "Unknown";
+    }
+  },
+  [](auto) { return std::string{}; })
+
+NVBENCH_DECLARE_ENUM_TYPE_STRINGS(
+  cudf_io::io_type,
+  [](auto value) {
+    switch (value) {
+      case cudf_io::io_type::FILEPATH: return "FILEPATH";
+      case cudf_io::io_type::HOST_BUFFER: return "HOST_BUFFER";
+      default: return "Unknown";
+    }
+  },
+  [](auto) { return std::string{}; })
+
+NVBENCH_DECLARE_ENUM_TYPE_STRINGS(
+  cudf_io::compression_type,
+  [](auto value) {
+    switch (value) {
+      case cudf_io::compression_type::SNAPPY: return "SNAPPY";
+      case cudf_io::compression_type::NONE: return "NONE";
+      default: return "Unknown";
+    }
+  },
+  [](auto) { return std::string{}; })
+
+void orc_read_common(cudf_io::orc_writer_options const& opts,
+                     cuio_source_sink_pair& source_sink,
+                     nvbench::state& state)
+{
+  cudf_io::write_orc(opts);
+
+  cudf_io::orc_reader_options read_opts =
+    cudf_io::orc_reader_options::builder(source_sink.make_source_info());
+
+  auto mem_stats_logger = cudf::memory_stats_logger();  // init stats logger
+  state.set_cuda_stream(nvbench::make_cuda_stream_view(cudf::default_stream_value.value()));
+  state.exec(nvbench::exec_tag::sync | nvbench::exec_tag::timer,
+             [&](nvbench::launch& launch, auto& timer) {
+               try_drop_l3_cache();
+
+               timer.start();
+               cudf_io::read_orc(read_opts);
+               timer.stop();
+             });
+
+  auto const time = state.get_summary("nv/cold/time/gpu/mean").get_float64("value");
+  state.add_element_count(static_cast<double>(data_size) / time, "bytes_per_second");
+  state.add_element_count(mem_stats_logger.peak_memory_usage(), "peak_memory_usage");
+  state.add_element_count(source_sink.size(), "encoded_file_size");
```

Comment:
This use of `add_element_count` surprises me. We're basically using it to store arbitrary integers, stretching its core purpose a bit. 

Also what do you think about adding `state.add_global_memory_writes` to show the size of the file in memory?

Reply:
> We're basically using it to store arbitrary integers, stretching its core purpose a bit.

Right, I see your concern. The latter two reporting peak memory usage and encoded file size are still fine but an API like `add_element_rate` would be more proper for mesurement like bytes per second.

> Also what do you think about adding state.add_global_memory_writes to show the size of the file in memory?

I spent some time looking into auto throughput metrics like [collect_stores_efficiency](https://github.com/NVIDIA/nvbench/blob/12d13bdc5e74801645eba3e46a64081b9b020dca/nvbench/state.cuh#L242) but found it's not easy to use since running it requires the system have performance counter privileges. e.g.
```bash
Run:  [1/8] throughput_bench [Device=0 T=1 Stride=1]
Warn: CUPTI failed to construct profiler: /home/yunsongw/Work/nvbench/nvbench/cupti_profiler.cxx:48: CUPTI call returned error: CUPTI_ERROR_INSUFFICIENT_PRIVILEGES
Fail: Unexpected error: /home/yunsongw/Work/nvbench/nvbench/cupti_profiler.cxx:48: CUPTI call returned error: CUPTI_ERROR_INSUFFICIENT_PRIVILEGES
Run:  [2/8] throughput_bench [Device=0 T=1 Stride=4]
Warn: CUPTI failed to construct profiler: /home/yunsongw/Work/nvbench/nvbench/cupti_profiler.cxx:48: CUPTI call returned error: CUPTI_ERROR_INSUFFICIENT_PRIVILEGES
Fail: Unexpected error: /home/yunsongw/Work/nvbench/nvbench/cupti_profiler.cxx:48: CUPTI call returned error: CUPTI_ERROR_INSUFFICIENT_PRIVILEGES
```
`bytes_per_second` measurement already includes the information provided by `add_global_memory_writes` so adding it won't help much in this particular case.